### PR TITLE
Fix nvcc math functions for MSVC 2019

### LINKED ISSUE #31108: test_int_pow_cuda failed on Windows
## To Reproduce

Run the following code:
```python
import torch

# The following ones will fail.
a = torch.tensor([-1,  1], device='cuda:0', dtype=torch.int64).pow(4); print(a)
a = torch.tensor([-1,  1], device='cuda:0', dtype=torch.int32).pow(4); print(a)
a = torch.tensor([-1,  1], device='cuda:0', dtype=torch.int16).pow(4); print(a)
a = torch.tensor([-1,  1], device='cuda:0', dtype=torch.int8).pow(4); print(a)

# The following ones will be okay.
a = torch.tensor([-1,  1], device='cuda:0', dtype=torch.int64).pow(2); print(a)
a = torch.tensor([-1,  1], device='cuda:0', dtype=torch.int32).pow(2); print(a)
a = torch.tensor([-1,  1], device='cuda:0', dtype=torch.int16).pow(2); print(a)
a = torch.tensor([-1,  1], device='cuda:0', dtype=torch.int8).pow(2); print(a)
```

Error message:
```
  File "<string>", line 1, in <module>
  File "C:\Users\circleci\project\build\win_tmp\build\torch\tensor.py", line 162, in __repr__
    return torch._tensor_str._str(self)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\_tensor_str.py", line 311, in _str
    tensor_str = _tensor_str(self, indent)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\_tensor_str.py", line 209, in _tensor_str
    formatter = _Formatter(get_summarized_data(self) if summarize else self)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\_tensor_str.py", line 83, in __init__
    value_str = '{}'.format(value)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\tensor.py", line 412, in __format__
    return self.item().__format__(format_spec)
RuntimeError: CUDA error: unspecified launch failure
```

## Expected behavior

No error,

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A

OS: Microsoft Windows Server 2019 Datacenter
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.7
Is CUDA available: N/A
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 426.00
cuDNN version: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\bin\cudnn64_7.dll

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect

## Additional context

I'm pretty sure that TDR is not a problem because the GPU is already in TCC mode.
BTW, it is not a problem in 1.3.1 too.

cc @ngimel @peterjc123

### PR DESCRIPTION
**Pull request by peterjc123**
Fixes https://github.com/pytorch/pytorch/issues/31108.

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 8759e2ad:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=8759e2ad91cd06ad3ad4243f184910a92ed27412).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4086232) caffe2_onnx_py3_6_clang7_ubuntu16_04_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=78773729) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=78773729))
<details>
<summary>
<code>ore dumped) pytest "${args[@]}" -k 'not (TestOperators and test_full_like) and not (TestOperators and test_zeros_like) and not (TestOperators and test_ones_like) and not (TestModels and test_vgg16) and not (TestModels and test_vgg16_bn) and not (TestModels and test_vgg19) and not (TestModels and test_vgg19_bn)' --ignore "$top_dir/test/onnx/test_pytorch_onnx_onnxruntime.py" --ignore "$top_dir/test/onnx/test_custom_ops.py" --ignore "$top_dir/test/onnx/test_models_onnxruntime.py" "${test_paths[@]}"</code>
</summary>

```
Dec 30 16:01:47   File "/usr/local/lib/python3.6/dist-packages/pluggy/hooks.py", line 286 in __call__ 
Dec 30 16:01:47   File "/usr/local/lib/python3.6/dist-packages/_pytest/main.py", line 237 in _main 
Dec 30 16:01:47   File "/usr/local/lib/python3.6/dist-packages/_pytest/main.py", line 193 in wrap_session 
Dec 30 16:01:47   File "/usr/local/lib/python3.6/dist-packages/_pytest/main.py", line 230 in pytest_cmdline_main 
Dec 30 16:01:47   File "/usr/local/lib/python3.6/dist-packages/pluggy/callers.py", line 187 in _multicall 
Dec 30 16:01:47   File "/usr/local/lib/python3.6/dist-packages/pluggy/manager.py", line 86 in <lambda> 
Dec 30 16:01:47   File "/usr/local/lib/python3.6/dist-packages/pluggy/manager.py", line 92 in _hookexec 
Dec 30 16:01:47   File "/usr/local/lib/python3.6/dist-packages/pluggy/hooks.py", line 286 in __call__ 
Dec 30 16:01:47   File "/usr/local/lib/python3.6/dist-packages/_pytest/config/__init__.py", line 90 in main 
Dec 30 16:01:47   File "/usr/local/bin/pytest", line 11 in <module> 
re dumped) pytest "${args[@]}" -k 'not (TestOperators and test_full_like) and not (TestOperators and test_zeros_like) and not (TestOperators and test_ones_like) and not (TestModels and test_vgg16) and not (TestModels and test_vgg16_bn) and not (TestModels and test_vgg19) and not (TestModels and test_vgg19_bn)' --ignore "$top_dir/test/onnx/test_pytorch_onnx_onnxruntime.py" --ignore "$top_dir/test/onnx/test_custom_ops.py" --ignore "$top_dir/test/onnx/test_models_onnxruntime.py" "${test_paths[@]}" 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: aten/src/ATen/native/cuda/PowKernel.cu [hunk 1]
```diff
@@ -22,16 +22,88 @@ static inline __host__ __device__ T powi(T a, T b) {
...
+// applied to the result of the inline function, and thus the result is incorrect.
+//   e.g. if we use 1.0 / sqrt(2) for 2 ^ (-0.5) in MSVC, we get
+//          int(2 ^ (-0.5)) = int(1.0 / sqrt(2)) = int(1.0 / int(1.414)) = int(1.0 / 1) = 1
+//        However, the correct result is 
+//          int(2 ^ (-0.5)) = int(1.0 / 1.414) = 0
+#ifdef _MSC_VER
+// Functions for pow
+// pow for at::Half
+static inline __host__ __device__ at::Half pow_(at::Half base, at::Half exp) {
+  return static_cast<at::Half>(std::pow(static_cast<double>(base), static_cast<double>(exp)));
```
Comment (gchanan) [inline aten/src/ATen/native/cuda/PowKernel.cu]: why do we go to double for float16 precision instead of float32?

Comment (peterjc123) [reply aten/src/ATen/native/cuda/PowKernel.cu]: Let me fix that.

**Review by ngimel — APPROVED** [review]
Thanks, LGTM! I'm glad this is solved.

**Comment by peterjc123** [comment]
I have no idea why the rocm builds are failing.

**Comment by peterjc123** [comment]
cc @bddppq @iotamudelta

**Comment by bddppq** [comment]
@peterjc123 Yeah ROCm doeesn't support putting `std::` functions in device code

**Pull request closed without merging**